Add README for custom pylint plugins (#170465)
home-assistant/core: pylint/plugins/README.md

inference, cross-file analysis (reading `manifest.json`, `quality_scale.yaml`),
or AST patterns that go beyond what a linter operating on a single file
can detect.

# Setup

This plugin is designed for internal use by Home Assistant Core's CI/CD
pipeline. It is not intended for external use, such as linting custom
integration repositories.

Plugins are loaded via the `load-plugins` setting in `pyproject.toml`.
The `init-hook` adds `pylint/plugins` to `sys.path` so these modules
are importable.

# Why not (just) Ruff?

Home Assistant uses both [Ruff](https://docs.astral.sh/ruff/) and pylint.
Ruff handles fast, single-file linting (import sorting, formatting, common
Python issues). These pylint checkers cover patterns that Ruff cannot:

- **Cross-file analysis**: reading `manifest.json` to check
  `integration_type`, reading `quality_scale.yaml` to verify claims.
- **Type inference**: resolving decorator names like
  `_pytest.fixtures.FixtureFunctionMarker` through imports.
- **Complex AST patterns**: tracing variable assignments to API calls
  (e.g., detecting that a variable assigned from `data[CONF_HOST]` later
  flows into `async_set_unique_id()`).

> # Disabling checks

Reviewer: since these "Disabling checks" instructions are generic and not specific to our pylint plugins, I would move this out into our dev docs repo

Author: Yeah I think it might be good to add them there as well 👍

Logger messages must start with a capital letter. This ensures consistency
across all integrations.

## `hass_imports` checker

Enforces import conventions for Home Assistant integrations. Integrations
should use relative imports for their own modules and follow specific
patterns for cross-integration references.

### `W7421`: `hass-relative-import`

Use relative imports within an integration (e.g., `from .const import DOMAIN`
instead of `from homeassistant.components.myintegration.const import DOMAIN`).

### `W7422`: `hass-deprecated-import`

Import uses a deprecated path that has been moved or renamed.

### `W7423`: `hass-absolute-import`

Use absolute imports when referencing modules outside the current integration.

### `W7424`: `hass-component-root-import`

Do not import from another integration's internal modules. Only import from
the integration's top-level public API.

### `W7425`: `hass-helper-namespace-import`

Use the helper namespace import pattern for helper modules.

### `W7426`: `hass-import-constant-alias`

Aliased `DOMAIN` import from another integration should use a descriptive alias.

### `W7427`: `hass-import-constant-unnecessary-alias`

> Unnecessary alias for `DOMAIN` import -- the alias matches the original name.

Reviewer: > the alias matches the original name.

this last part is confusing to me.

```suggestion
Unnecessary alias when importing DOMAIN from within the same integration.
```

signatures defined by Home Assistant's platform interfaces.

### `W7431`: `hass-argument-type`

Function argument should have the specified type hint. Platform functions
like `async_setup_entry` have well-defined signatures that must be followed.

### `W7432`: `hass-return-type`

Function should have the specified return type hint.

### `W7433`: `hass-consider-usefixtures-decorator`

Test function should use `@pytest.mark.usefixtures("fixture_name")` instead
of accepting an unused fixture as a parameter.

## `hass_decorator` checker

Validates decorator usage on functions and fixtures.

### `W7471`: `hass-async-callback-decorator`

A coroutine function (`async def`) should not be decorated with `@callback`.
The `@callback` decorator is only for synchronous functions that should be
called from the event loop without scheduling.

### `W7472`: `hass-pytest-fixture-decorator`

Pytest fixture has invalid scope or autouse configuration. For example,
`session`-scoped fixtures in component tests should use `package` scope
or lower.

## `hass_inheritance` checker

Validates entity class inheritance chains.

### `W7411`: `hass-invalid-inheritance`

> Entity class has an invalid inheritance chain. Entity classes must properly

Reviewer: this makes it sound like the checker enforces a broad inheritance discipline, when in reality it's narrowly scoped. Can we make it more specific?

Author: Done
